mxfusion/inference/dist_grad_based_inference.py

```diff
@@ -0,0 +1,161 @@
+#   permissions and limitations under the License.
+# ==============================================================================
+
+
+from .dist_batch_loop import DistributedBatchInferenceLoop
+from .inference import Inference
+from .dist_minibatch_loop import DistributedMinibatchInferenceLoop
+from ..util.inference import discover_shape_constants, init_outcomes
+
+
+class DistributedGradBasedInference(Inference):
+    """
+    The abstract class for distributed gradient-based inference methods.
+    An inference method consists of a few components: the applied inference algorithm, the model definition
+    (optionally a definition of posterior approximation), the inference parameters.
+
+    :param inference_algorithm: The applied inference algorithm
+    :type inference_algorithm: InferenceAlgorithm
+    :param grad_loop: The reference to the main loop of gradient optimization
+    :type grad_loop: GradLoop
+    :param constants: Specify a list of model variables as constants
+    :type constants: {Variable: mxnet.ndarray}
+    :param hybridize: Whether to hybridize the MXNet Gluon block of the inference method.
+    :type hybridize: boolean
+    :param dtype: data type for internal numerical representation
+    :type dtype: {numpy.float64, numpy.float32, 'float64', 'float32'}
+    :param context: The MXNet context
+    :type context: {mxnet.cpu or mxnet.gpu}
+    :param logger: The logger to send logs to
+    :type logger: :class:`inference.Logger`
+    """
+
+    def __init__(self, inference_algorithm, grad_loop=None, constants=None,
+                 hybridize=False, dtype=None, context=None, logger=None):
+        if grad_loop is None:
+            grad_loop = DistributedBatchInferenceLoop()
+        super(DistributedGradBasedInference, self).__init__(
+            inference_algorithm=inference_algorithm, constants=constants,
+            hybridize=hybridize, dtype=dtype, context=context, logger=logger)
+        self._grad_loop = grad_loop
+
+    def create_executor(self):
+        """
+        Return a MXNet Gluon block responsible for the execution of the inference method.
+        """
+        from mxfusion.inference import StochasticVariationalInference
+
+        rv_scaling = {}
+        if isinstance(self.inference_algorithm, StochasticVariationalInference):
+            import horovod.mxnet as hvd
+
+            for _, variable in enumerate(self.inference_algorithm.model.get_latent_variables(self.inference_algorithm.observed)):
+                rv_scaling[variable] = 1/hvd.size()
+            print(rv_scaling)
+
+            if rv_scaling == {}:
+                rv_scaling = None
+
+            rv_scaling = {v.uuid: s for v, s in rv_scaling.items()} \
+                if rv_scaling is not None else rv_scaling
+
+        if isinstance(self._grad_loop, DistributedMinibatchInferenceLoop):
+            # rv_scaling = self._grad_loop.rv_scaling
+            if(self._grad_loop.rv_scaling is not None):
```

Comment:
I don't understand why you store the rv_scaling in self._grad_loop.rv_scaling only if its MiniBatch. What is the reason for having different behavior there?